### PR TITLE
fix: cw pool import export genesis

### PR DESCRIPTION
**Pull request by czarcas7ic**
Closes: #XXX

## What is the purpose of the change

Export and import cwpools in genesis

### CODE REVIEW

File: x/cosmwasmpool/module/module.go [hunk 1]
```diff
@@ -30,7 +30,9 @@ var (
 	_ module.AppModuleBasic = AppModuleBasic{}
 )
 
-type AppModuleBasic struct{}
+type AppModuleBasic struct {
+	cdc codec.Codec
```
Comment (p0mvn) [inline x/cosmwasmpool/module/module.go:34]: Where is this being set?

Comment (czarcas7ic) [reply x/cosmwasmpool/module/module.go:34]: Appreciate you catching this, set them here https://github.com/osmosis-labs/osmosis/pull/5696/commits/1d4505c2b1501111efc3019cb870b7d19e316322

**Review by p0mvn — APPROVED** [review]
LGTM. Thanks for adding this

Consider adding a simple test to make sure everything works as intended

File: x/cosmwasmpool/genesis.go [hunk 2]
```diff
@@ -1,20 +1,54 @@
...
 	k.SetParams(ctx, gen.Params)
+
+	// Add each genesis state pool to the x/cosmwasmpool module's state
+	for _, any := range gen.Pools {
+		var pool types.CosmWasmExtension
+		err := unpacker.UnpackAny(any, &pool)
+		if err != nil {
+			panic(err)
+		}
+		pool.SetWasmKeeper(k.wasmKeeper)
```
Comment (p0mvn) [inline x/cosmwasmpool/genesis.go:21]: Actually, why is this needed? We should be setting the store model to state, not the wrapped that has `wasmKeeper`

File: x/cosmwasmpool/store.go [hunk 3]
```diff
@@ -12,9 +14,15 @@ import (
...
+func (k Keeper) SetPool(ctx sdk.Context, pool types.CosmWasmExtension) error {
+	poolModel, ok := pool.(*model.Pool)
+	if !ok {
+		return errors.New("invalid pool type when setting cosmwasm pool")
+	}
 	store := ctx.KVStore(k.storeKey)
-	osmoutils.MustSet(store, types.FormatPoolsPrefix(pool.GetId()), pool.GetStoreModel())
+	key := types.FormatPoolsPrefix(pool.GetId())
+	osmoutils.MustSet(store, key, poolModel)
+	return nil
```
Comment (p0mvn) [inline x/cosmwasmpool/store.go:25]: Why do we need these changes? Old logic with `pool.GetStoreModel()` looks correct to me at first glance.

The store model is what needs to be serialized to state, not the wrapper with `wasmKeeper`.

Also, by casting the input to `model.Pool`, this essentially becomes a store model with `wasmKeeper` stripped away

File: x/cosmwasmpool/pool_module.go [hunk 4]
```diff
@@ -99,12 +102,15 @@ func (k Keeper) GetPool(ctx sdk.Context, poolId uint64) (poolmanagertypes.PoolI,
...
+			pool := model.CosmWasmPool{}
 			err := k.cdc.Unmarshal(value, &pool)
 			if err != nil {
 				return nil, err
 			}
-			return &pool, nil
+			return &model.Pool{
+				CosmWasmPool: pool,
+				WasmKeeper:   k.wasmKeeper,
+			}, nil
```
Comment (p0mvn) [inline x/cosmwasmpool/pool_module.go:113]: I don't follow this change either. We should be careful because this might break CLI serialization.

The reason why we return store model and not the wrapper `model.Pool` is because `model.Pool` is a pure go struct. It cannot be properly serialized or deserialized. As a result, this would break CLI and queries.

For the purposes of import and export genesis, we should be fine operating on `model.CosmwasmPool`. There is no need for `model.Pool`. It is only needed, when we want to run certain `PoolI` methods that require making calls to the contract

File: x/cosmwasmpool/genesis_test.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	// err := simapp.FundAccount(s.App.BankKeeper, s.Ctx, acc1, sdk.NewCoins(
+	// 	sdk.NewCoin("uosmo", sdk.NewInt(10000000000)),
+	// 	sdk.NewInt64Coin("foo", 100000),
+	// 	sdk.NewInt64Coin("bar", 100000),
+	// ))
+	// s.Require().NoError(err)
+
+	// msg := model.NewMsgCreateCosmWasmPool(validTransmuterCodeId, s.TestAccs[0], s.GetDefaultTransmuterInstantiateMsgBytes())
+	// _, err = s.App.PoolManagerKeeper.CreatePool(s.Ctx, msg)
+	// s.Require().NoError(err)
```
Comment (p0mvn) [inline x/cosmwasmpool/genesis_test.go:115]: nit: dead code

File: x/cosmwasmpool/genesis_test.go [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+	s.FundAcc(s.TestAccs[0], initalDefaultSupply)
+	s.PrepareCustomTransmuterPool(s.TestAccs[0], defaultDenoms)
+
+	// msg = model.NewMsgCreateCosmWasmPool(validTransmuterCodeId, s.TestAccs[0], s.GetDefaultTransmuterInstantiateMsgBytes())
+	// _, err = s.App.PoolManagerKeeper.CreatePool(s.Ctx, msg)
+	// s.Require().NoError(err)
```
Comment (p0mvn) [inline x/cosmwasmpool/genesis_test.go:121]: and here

File: x/cosmwasmpool/genesis_test.go [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+
+	fmt.Println("pools liq", pools[0].(types.CosmWasmExtension).GetTotalPoolLiquidity(s.Ctx))
+
+	liquiditytotal, err := s.App.CosmwasmPoolKeeper.GetTotalLiquidity(s.Ctx)
+	s.Require().NoError(err)
+	fmt.Println("TOTAL PRE liquidity", liquiditytotal.String())
+
+	// We need to export the wasm module and reimport it after we reset the test environment
+	// This is because cosmwasmpools point to a contract address, and if we don't export this as well
+	// this test will fail.
```
Comment (p0mvn) [inline x/cosmwasmpool/genesis_test.go:67]: Why? I don't see why we need to export / import `model.Pool` that has the `wasmKeeper`

**Review by p0mvn — CHANGES_REQUESTED** [review]

File: x/cosmwasmpool/genesis_test.go (hunk 6 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/genesis_test.go]: https://github.com/osmosis-labs/osmosis/pull/5696/commits/a7590c0e1a8523d2d582528b1e31982d00362f12

File: x/cosmwasmpool/genesis_test.go (hunk 5 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/genesis_test.go]: https://github.com/osmosis-labs/osmosis/pull/5696/commits/a7590c0e1a8523d2d582528b1e31982d00362f12

File: x/cosmwasmpool/genesis.go (hunk 2 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/genesis.go:21]: We get the following error in the initiGenesis test when querying the spreadFactor without this line:

```
    --- FAIL: TestPoolModuleSuite/TestInitGenesis (0.40s)
        /Users/czar/Documents/osmosis/x/cosmwasmpool/suite.go:87: test panicked: kv store with key KVStoreKey{0x140010fbdc0, wasm} has not been registered in stores
            goroutine 56 [running]:
            runtime/debug.Stack()
            	/usr/local/go/src/runtime/debug/stack.go:24 +0x64
            github.com/stretchr/testify/suite.failOnPanic(0x140000ca4e0, {0x1021dbe80, 0x14000eaddd0})
            	/Users/czar/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:87 +0x34
            github.com/stretchr/testify/suite.Run.func1.1()
            	/Users/czar/go/pkg/mod/github.com/stretchr/testify@v1.8.4/suite/suite.go:183 +0x1f4
            panic({0x1021dbe80, 0x14000eaddd0})
            	/usr/local/go/src/runtime/panic.go:884 +0x204
            github.com/cosmos/cosmos-sdk/store/cachemulti.Store.GetKVStore({{0x1025b86b0, 0x14000f0d530}, 0x14000f0d5c0, 0x14000f22450, {0x0, 0x0}, 0x0, 0x14000f224e0}, {0x102588708, 0x140010fbdc0})
```

File: x/cosmwasmpool/store.go (hunk 3 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/store.go]: This is an unnecessary change, was just done in the debugging process. Reverted https://github.com/osmosis-labs/osmosis/pull/5696/commits/a7590c0e1a8523d2d582528b1e31982d00362f12

File: x/cosmwasmpool/pool_module.go (hunk 4 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/pool_module.go:110]: When removing this change, the following serialization fails, where we need to utilize cosmwasmPool in order to call GetTotalPoolLiquidity

https://github.com/osmosis-labs/osmosis/blob/79b4a8717ec361dcea779ee0360324279e7e806c/x/cosmwasmpool/pool_module.go#L353-L358

File: x/cosmwasmpool/genesis_test.go (hunk 7 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/genesis_test.go:46]: Moving this over from DMs:

Without export / import genesis (and calling the GetPoolById) I still get the error when trying to GetSpreadFactor:

```
--- FAIL: TestPoolModuleSuite/TestInitGenesis (0.24s)
       /osmosis/x/cosmwasmpool/suite.go:87: test panicked: contract: not found [osmosis-labs/wasmd@v0.31.0-osmo-v16/x/wasm/keeper/keeper.go:714]
            goroutine 10 [running]:
            runtime/debug.Stack()
```

File: x/cosmwasmpool/genesis.go (hunk 2 above)
Comment (czarcas7ic) [reply x/cosmwasmpool/genesis.go]: Added your recommendation which fixed it here https://github.com/osmosis-labs/osmosis/pull/5696/commits/a4bafb5d258769bf7eb6f871c7ec3de6dd0d8430

File: x/cosmwasmpool/pool_module.go (hunk 4 above)
Comment (p0mvn) [reply x/cosmwasmpool/pool_module.go:110]: This change will break the `AllPools` `x/poolmanager` query and CLI because `model.Pool` cannot be serialized.

Please test to make sure. If that's the case, please implement new logic for the `GetTotalPoolLiquidity`

Comment (czarcas7ic) [reply x/cosmwasmpool/pool_module.go:110]: @p0mvn should I just make a GetPoolsWasmExtension and only use that when exporting and importing genesis?

Comment (p0mvn) [reply x/cosmwasmpool/pool_module.go:110]: Yes, that makes sense. However, both `model.Pool` and `model.CosmWasmPool` implement `types.CosmWasmExtension` so the name doesn't seem to convey what we would like. We would like to differentiate between the `model.Pool` that has wasm keeper and `model.CosmWasmPool` that does not.

So I suggest considering `GetPoolsWithWasmKeeper`  or something similar

File: x/cosmwasmpool/genesis_test.go (hunk 7 above)
Comment (p0mvn) [reply x/cosmwasmpool/genesis_test.go:50]: Discussed offline. This makes sense to me now

**Review by p0mvn — APPROVED** [review]
LGTM once the last thread is resolved. Nice work!

File: x/cosmwasmpool/genesis.go [hunk 8]
```diff
@@ -26,7 +26,7 @@ func (k *Keeper) InitGenesis(ctx sdk.Context, gen *types.GenesisState, unpacker
 func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
 	params := k.GetParams(ctx)
 
-	pools, err := k.GetPools(ctx)
+	pools, err := k.GetPoolsWithWasmKeeper(ctx)
```
Comment (p0mvn) [inline x/cosmwasmpool/genesis.go]: This should be `GetPools`, no reason to export the one with wasm keeper

Comment (czarcas7ic) [reply x/cosmwasmpool/genesis.go]: Same error occurs if I do not do this

Comment (czarcas7ic) [reply x/cosmwasmpool/genesis.go]: Disregard, was maybe cache or something, this works
